icc-x-api/icc-crypto-x-api.ts

```diff
@@ -154,24 +154,134 @@ export class IccCryptoXApi {
-      const pk = exportedKey as ArrayBuffer
-      const shares = this._shamir.share(this._utils.ua2hex(pk), notaries.length, threshold)
-      return Promise.all(
-        notaries.map((notary, idx) => {
-          const notaryPubKey = utils.spkiToJwk(utils.hex2ua(notary.publicKey!))
-          return this._RSA
-            .importKey("jwk", notaryPubKey, ["encrypt"])
-            .then(key => {
-              this._RSA.encrypt(key, this._utils.hex2ua(shares[idx]))
+    threshold?: number
+  ): Promise<HealthcarePartyDto> {
+    return this._RSA.loadKeyPairImported(hcp.id!).then(keyPair =>
+      this._RSA.exportKey(keyPair.privateKey, "pkcs8").then(exportedKey => {
+        const privateKey = exportedKey as ArrayBuffer
+        const nLen = notaries.length
+        const shares =
+          nLen == 1
+            ? [privateKey]
+            : this._shamir
+                .share(this._utils.ua2hex(privateKey), nLen, threshold || nLen)
+                .map(share => this.utils.hex2ua(share))
+
+        return _.reduce(
+          notaries,
+          (queue, notary, idx) => {
+            return queue.then(async hcp => {
+              const hcParty = hcp.hcPartyKeys![notary.id!]
+                ? hcp
+                : ((await this.generateKeyForDelegate(hcp.id!, notary.id!)) as HealthcarePartyDto)
+
+              try {
+                const importedAESHcPartyKey = await this.decryptHcPartyKey(
+                  hcParty.id!,
+                  notary.id!,
+                  hcParty.hcPartyKeys![notary.id!][1],
+                  false
+                )
+                const encryptedShamirPartition = await this.AES.encrypt(
+                  importedAESHcPartyKey.key,
+                  shares[idx]
+                )
+
+                hcParty.privateKeyShamirPartitions = hcParty.privateKeyShamirPartitions || {}
+                hcParty.privateKeyShamirPartitions[notary.id!] = this.utils.ua2hex(
+                  encryptedShamirPartition
+                )
+              } catch (e) {
+                console.log("Error during encryptedShamirRSAKey", notary.id, e)
+              }
+              return hcParty
             })
-            .then(k => [notary.id, k])
-        })
-      ).then(keys => _.fromPairs(keys) as Map<string, string>)
-    })
+          },
+          Promise.resolve(hcp)
+        )
+      })
+    )
+  }
+
+  /* Reconstructs the hcp's private key from the notaries' shamir shares and stores it in localstorage.
+  The retrieval procedure of the shares is not designed or implemented yet.  Therefore, it currently only
+  works if the private key of the notaries are stored in local storage (e.g. notaries = [hcp parent]).
+   * @param hcp : the hcp whose key we want to reconstruct
+   * @param notaries : holders of the shamir shares
+  **/
+  async decryptedShamirRSAKey(
+    hcp: HealthcarePartyDto,
+    notaries: Array<HealthcarePartyDto>
+  ): Promise<void> {
```

Comment:
The name of the method should be more explicit regarding it purpose which is very specific.